Missing IList API

emilypi opened the pull request:
Sup folks, here's a basic API for `IList`

emilypi commented:
@xuwei-k thanks for tagging!

jdegoes commented on base/shared/src/main/scala/scalaz/data/ilist.scala:
```diff
@@ -37,6 +41,87 @@ trait IListModule {
+      IList.uncons(self) match {
+        case Maybe2.Empty2()     => Maybe.empty
+        case Maybe2.Just2(_, as) => Maybe.just(as)
+      }
```
Could we call these `head` and `tail`?

emilypi replied:
sure - i'm on board with making the safe versions mandatory 😄

jdegoes commented on base/shared/src/main/scala/scalaz/data/ilist.scala:
```diff
@@ -37,6 +41,87 @@ trait IListModule {
+    }
+
+    def size: Int =
+      IList.foldLeft[A, Int]((b, a) => 1 + b, 0, self)
```
This will be faster specialized.

emilypi replied:
Good call!

alexknvl replied:
He means that writing it out in terms of pattern matching + `@tailrec` function will be faster, because `Int`s won't be boxed and the function will be "inlined" (manually).

emilypi replied:
oh, sure i can do that. I thought he was referring to `@specialized` and I was like 😖

emilypi commented:
@alexknvl would you be okay with renaming `empty` to `nil`? We have a naming conflict with `Maybe2#empty` in `AllFunctions`, and i'm sure we'd like to inline our `empty`'s for `IList`

emilypi commented:
@jdegoes I've addressed `size` as a special case and added a few more standard methods to the bunch. I think this is relatively complete as APIs go. I'm leaving the rest of the functionality up to the typeclasses.
